pageserver/src/object_repository.rs

```diff
@@ -838,6 +809,33 @@ impl ObjectTimeline {
             ancestor_lsn: self.ancestor_lsn,
         })
     }
+
+    //
+    // Helper functions to store different kinds of objects to the underlying ObjectStore
+    //
+    fn put_page_entry(&self, tag: &BufferTag, lsn: Lsn, val: PageEntry) -> Result<()> {
+        let key = ObjectKey {
+            timeline: self.timelineid,
+            tag: ObjectTag::RelationBuffer(*tag),
+        };
+        let val = ObjectValue::Page(val);
+
+        self.obj_store.put(&key, lsn, &ObjectValue::ser(&val)?)
```

Comment:
I thought about that too, but it would require making `ObjectValue` public. Currently, `ObjectStore` doesn't know anything about the values that it stores. It would be nice if we could pass an `ObjectValue` to the functions in `ObjetStore`, but keep it opaque so that ObjectStore could only access the serialize/deserialize functions on it.

Reply:
There are two contradictory directions:
1. Make our storage API as simple as possible and implement most of the logic at top level. So  storage is just KV database, with opaque K and V. Well, we need to enforce some ordering of K, so it is not so opaque. It makes it possible to provide alternative storage implementation with minimal code duplication.
2. Make it really "object" storage and let it know about format and may be even semantic of stored keys/values. It allows to make some storage-specific optimizations. For example implement specialized garbage collector or even eliminate garbage collection of our storage format doesn't require it. 

We have made several steps in opposite directions in this axis. Originally GC is done by storage, now it is moved to upper level. Currently object storage is aware about key, but value is opaque. I do not like this asymmetry, but not sure that moving serialization of value to storage layer is better solution.

In any case, I think that we should not try now to design ideal  object storage API. First of all Eric is working on new storage format it may affect object storage API. Also it is very hard to design good abstract API without having several its realizations. Once we have our own object storage implementation (as alternative to RockDB), it will be easier to understand  how to distribute responsibilities between object repository and object storage.